File: app/src/main/java/de/prttstft/materialmensa/model/Meal.java
package de.prttstft.materialmensa.model;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("unused")
public class Meal {
    @SerializedName("date") private String date;
    @SerializedName("name_de") private String nameDe;
    @SerializedName("name_en") private String nameEn;
    @SerializedName("description_de") private String descriptionDe;
    @SerializedName("description_en") private String descriptionEn;
    @SerializedName("category_de") private String categoryDe;
    @SerializedName("category_en") private String categoryEn;
    @SerializedName("subcategory") private String subcategory;
    @SerializedName("subcategory_en") private String subcategoryEn;
    @SerializedName("priceStudents") private float priceStudents;
    @SerializedName("priceWorkers") private float priceWorkers;
    @SerializedName("priceGuests") private float priceGuests;
    @SerializedName("allergens") private List<String> allergens = new ArrayList<>();
    @SerializedName("order_info") private int orderInfo;
    @SerializedName("badges") private List<String> badges = new ArrayList<>();
    @SerializedName("restaurant") private String restaurant;
    @SerializedName("pricetype") private String pricetype;
    @SerializedName("image") private String image;
    @SerializedName("thumbnail") private String thumbnail;
    private String priceString;
    private int orderNumber;
    private String customDescription;

    // Getters & Setters

    public String getDate() {
        return date;
    }

    public String getNameDe() {
        return nameDe;
    }

    public String getNameEn() {
        return nameEn;
    }

    public String getDescriptionDe() {
        return descriptionDe;
    }

    public String getDescriptionEn() {
        return descriptionEn;
    }

    public String getCategoryDe() {
        return categoryDe;
    }

    public String getCategoryEn() {
        return categoryEn;
    }

    public String getSubcategory() {
        return subcategory;
    }

    public String getSubcategoryEn() {
        return subcategoryEn;
    }

    public float getPriceStudents() {
        return priceStudents;
    }

    public float getPriceWorkers() {
        return priceWorkers;
    }

    public float getPriceGuests() {
        return priceGuests;
    }

    public List<String> getAllergens() {
        return allergens;
    }

    public int getOrderInfo() {
        return orderInfo;
    }

    public List<String> getBadges() {
        return badges;
    }

    public String getRestaurant() {
        return restaurant;
    }

    public String getPricetype() {
        return pricetype;
    }

    public String getImage() {
        return image;
    }

    public String getThumbnail() {
        return thumbnail;
    }

    public String getPriceString() {
        return this.priceString;
    }

    public void setPriceString(String priceString) {
        this.priceString = priceString;
    }

    public int getOrderNumber() {
        return orderNumber;
    }

    public void setOrderNumber(int orderNumber) {
        this.orderNumber = orderNumber;
    }

    public String getCustomDescription() {
        return customDescription;
    }

    public void setCustomDescription(String customDescription) {
        this.customDescription = customDescription;
    }


    // Helpers

    public String getBadge() {
        if (badges != null && badges.size() > 0) {
            return badges.get(0);
        } else {
            return "";
        }
    }
}

File: app/src/main/java/de/prttstft/materialmensa/model/Meal.kt
@file:JvmName("Meal")
package de.prttstft.materialmensa.model

import com.google.gson.annotations.SerializedName
import java.util.*
import kotlin.comparisons.compareValuesBy

@SuppressWarnings("unused")
class Meal : Comparable<Meal> {

    @SerializedName("allergens") val allergens: List<String> = ArrayList()
    @SerializedName("badges") val badges: List<String> = ArrayList()
    @SerializedName("category") val category: String = ""
    @SerializedName("category_de") val categoryDe: String = ""
    @SerializedName("category_en") val categoryEn: String = ""
    @SerializedName("date") val date: String = ""
    @SerializedName("description_de") val descriptionDe: String = ""
    @SerializedName("description_en") val descriptionEn: String = ""
    @SerializedName("image") val image: String = ""
    @SerializedName("name_de") var nameDe: String = ""
    @SerializedName("name_en") var nameEn: String = ""
    @SerializedName("order_info") val orderInfo: Int = 0
    @SerializedName("priceGuests") val priceGuests: Float = 0.toFloat()
    @SerializedName("priceStudents") val priceStudents: Float = 0.toFloat()
    @SerializedName("priceWorkers") val priceWorkers: Float = 0.toFloat()
    @SerializedName("pricetype") val pricetype: String = ""
    @SerializedName("restaurant") val restaurant: String = ""
    @SerializedName("subcategory") val subcategory: String = ""
    @SerializedName("subcategory_en") val subcategoryEn: String = ""
    @SerializedName("thumbnail") val thumbnail: String = ""
    var customDescription: String = ""
    var downvotes: HashMap<String, Any> = HashMap()
    var isDownvoted: Boolean = false
    var isFiltered: Boolean = false
    var isUpvoted: Boolean = false
    var orderNumber: Int = 0
    var priceString: String = ""
    var score: Int = 0
    var upvotes: HashMap<String, Any> = HashMap()
    var hasScores: Boolean = false


    val badge: String
        get() {
            if (badges.size > 0) {
                return badges[0]
            } else {
                return ""
            }
        }

    override fun compareTo(other: Meal): Int {
        return compareValuesBy(this, other, { it.customSort() })
    }

    private fun customSort(): Int {
        if (isUpvoted) {
            return -10000000 - score
        }

        if (isDownvoted) {
            return +10000000 - score
        }

        if (score > 0) {
            return score * (-1)
        }

        if (score < 0) {
            return score * (-1)
        }

        return orderNumber
    }
}